pyccl/tests/test_cclobject.py

```diff
@@ -4,6 +4,25 @@
 import functools
 
 
+def test_fancy_repr():
+    # Test fancy-repr controls.
+    cosmo1 = ccl.CosmologyVanillaLCDM()
+    cosmo2 = ccl.CosmologyVanillaLCDM()
+
+    ccl.CCLObject._fancy_repr.disable()
+    assert "False" in repr(ccl.CCLObject._fancy_repr)
+    assert repr(cosmo1) == object.__repr__(cosmo1)
```

Comment:
Humm, since cosmo1 was created before disabling fancy_repr, shouldn't it have the fancy repr? Or do they only get created when you first do a comparison?

Reply:
First of all, this test is outdated, my `git` says "nothing to commit" and once again, github hasn't updated the changes.
Secondly, it turns out I was wrong yesterday when I told you on Slack that already-instantiated objects won't change. Every time `__repr__` is called, the interpreter re-calculates what the `__repr__` for that object should be, looking at its class. So if you change the class of an object, you also automatically change the object's method. Which means that it even works with instantiated objects.

Reply:
Okay, it looks like if you go to the "Files Changed" tab, it shows the correct (synced) file. So I can only assume that Github has forgot to put an `(outdated)` badge in the "Conversation" tab.

Reply:
OK, but if _repr is cached, it's not recalculated, right?

Reply:
It is. If `_repr` is cached, and in the meantime you disable and re-enable fancy-reprs, the `_repr` that was cached gets replaced (on enable):
```python
    def enable(cls):
        """Enable fancy representations if they exist."""
        for cl, method in cls._classes.items():
            FancyRepr.bind_and_replace(cl, method)
        cls._enabled = True
```
which calls
```python
    @classmethod
    def bind_and_replace(cls, cl, method):
        """Bind ``method`` to class ``cl``, and replace original with default.
        This helper only works for binding and replacing ``__repr__`` methods
        for ``CCLObjects``.
        """
        # If the class defines a custom `__repr__`, this will be the new
        # `_repr` (which is cached). Decorator `cached_property` requires
        # that `__set_name__` is called on it.
        bmethod = functools.cached_property(method)
        cl._repr = bmethod
        bmethod.__set_name__(cl, "_repr")
        # Fall back to using `__ccl_repr__` from `CCLObject`.
        cl.__repr__ = cl.__ccl_repr__
```

Reply:
See here for an example:
```python
import pyccl as ccl
cosmo = ccl.CosmologyVanillaLCDM()
cosmo.compute_linear_power()
pk = cosmo.get_linear_power()
print(pk)
ccl.CCLObject._fancy_repr.disable()
print(pk)
pk += 1
ccl.CCLObject._fancy_repr.enable()
print(pk)
```
The first print:
```
<pyccl.pk2d.Pk2D>
	+==============+=================================================================+
	| a \ log10(k) | -4.30e+00 -4.16e+00 -4.01e+00 ...  8.37e-01  9.29e-01  1.02e+00 |
	+==============+=================================================================+
	|  1.00e-02    |  1.84e-01  2.53e-01  3.49e-01 ...  1.09e-04  6.25e-05  3.56e-05 |
	|  1.26e-02    |  2.98e-01  4.11e-01  5.66e-01 ...  1.72e-04  9.83e-05  5.60e-05 |
	|  1.58e-02    |  4.82e-01  6.63e-01  9.13e-01 ...  2.71e-04  1.55e-04  8.81e-05 |
	|     ...      |                               ...                               |
	|  9.54e-01    |  1.19e+03  1.64e+03  2.26e+03 ...  6.06e-01  3.47e-01  1.97e-01 |
	|  9.77e-01    |  1.22e+03  1.68e+03  2.32e+03 ...  6.22e-01  3.56e-01  2.02e-01 |
	|  1.00e+00    |  1.25e+03  1.72e+03  2.37e+03 ...  6.37e-01  3.64e-01  2.07e-01 |
	+==============+=================================================================+
	| is_log = True , extrap_orders = (1, 2)                                         |
	| HASH_ARRS = 0x1b938ef61e5e1edb9                                                |
	+==============+=================================================================+
```

The second print
```
<pyccl.pk2d.Pk2D object at 0x7f6f21779450>
```

The third print
```
<pyccl.pk2d.Pk2D>
	+==============+=================================================================+
	| a \ log10(k) | -4.30e+00 -4.16e+00 -4.01e+00 ...  8.37e-01  9.29e-01  1.02e+00 |
	+==============+=================================================================+
	|  1.00e-02    |  1.18e+00  1.25e+00  1.35e+00 ...  1.00e+00  1.00e+00  1.00e+00 |
	|  1.26e-02    |  1.30e+00  1.41e+00  1.57e+00 ...  1.00e+00  1.00e+00  1.00e+00 |
	|  1.58e-02    |  1.48e+00  1.66e+00  1.91e+00 ...  1.00e+00  1.00e+00  1.00e+00 |
	|     ...      |                               ...                               |
	|  9.54e-01    |  1.19e+03  1.64e+03  2.26e+03 ...  1.61e+00  1.35e+00  1.20e+00 |
	|  9.77e-01    |  1.22e+03  1.68e+03  2.32e+03 ...  1.62e+00  1.36e+00  1.20e+00 |
	|  1.00e+00    |  1.25e+03  1.73e+03  2.37e+03 ...  1.64e+00  1.36e+00  1.21e+00 |
	+==============+=================================================================+
	| is_log = True , extrap_orders = (1, 2)                                         |
	| HASH_ARRS = 0x10759946ba24489cb                                                |
	+==============+=================================================================+
```

Reply:
So essentially by disabling we simply unlink `__repr__` from `_repr`. Note that `_repr` is the one that carries the workload when using fancy-reprs.

Reply:
right, but that's only if you change fancy_repr, if you don't, _repr is cached